File: libs/js/src/CustomNodesSorting.ts
import { Path } from './Path';
import { Node } from './Node';

class SortBucketList {
    private _path?: Path;
    private list: Array<Node>;

    constructor(path?: Path) {
        this._path = path;
        this.list = [];
    }

    public nodes() {
        return this.list;
    }

    public addNode(node: Node) {
        let found = false;
        for (let i = 0; i < this.list.length; i++)
        {
            var currentNode = this.list[i];
            let condition = false;
            if (this._path != null) {
                var a = node.pairsCount();
                var b = currentNode.pairsCount();
                if (a != b) {
                    condition = a < b;
                }
                else {
                    condition = node.value > currentNode.value;
                }
            }
            else {
                condition = node.value > currentNode.value;
            }

            if (condition) {
                this.list.splice(i, 0, node);
                found = true;
                break;
            }
        }

        // reached end of list without adding
        if (!found) {
            this.list.push(node);
        }
    }
}

export class CustomNodesSorting {
    private _path?: Path | null;
    private _maxN: number;

    constructor(path: Path | null, maxN: number) {
        this._path = path;
        this._maxN = maxN;
    }

    public sortNodes(nodes: Array<Node>){
        const sortList: Array<SortBucketList> = new Array(this._maxN + 1);

        for (let i = 0; i < nodes.length; i++) {
            var node = nodes[i];
            var pairsCount = this._path == null ? node.pairsCount() : node.pairsNotInPath(this._path);

            var list = sortList[pairsCount];
            if (list == null) {
                list = new SortBucketList(this._path);
            }

            list.addNode(node);

            sortList[pairsCount] = list;
        }

        let index = 0;
        for (let list of sortList) {
            if (list == null) {
                continue;
            }

            for(let node of list.nodes()) {
                nodes[index] = node;
                index+=1;
            }
        }
    }
}

File: libs/dotnet/SquareSums/CustomNodesSorting.cs
using System;
using System.Collections.Generic;

namespace SquareSums
{

    class SortBucketList
    {
        private readonly Path? _path;
        private readonly List<Node> _list;

        public SortBucketList(Path? path)
        {
            _path = path;
            _list = new List<Node>();
        }

        public IReadOnlyList<Node> Nodes => _list;

        public void AddNode(Node node)
        {
            bool found = false;
            for (int i = 0; i < _list.Count; i++)
            {
                var currentNode = _list[i];
                bool condition;
                if (_path != null)
                {
                    var a = node.PairsCount;
                    var b = currentNode.PairsCount;
                    if (a != b)
                    {
                        condition = a < b;
                    }
                    else
                    {
                        condition = node.Value > currentNode.Value;
                    }
                }
                else
                {
                    condition = node.Value > currentNode.Value;
                }

                if (condition)
                {
                    _list.Insert(i, node);
                    found = true;
                    break;
                }
            }

            // reached end of list without adding
            if (!found)
            {
                _list.Add(node);
            }
        }
    }
    
    public class CustomNodesSorting: INodesSorting
    {
        private readonly Path? _path;
        private readonly int _maxN;

        public CustomNodesSorting(Path? path, int maxN)
        {
            _path = path;
            _maxN = maxN;
        }
        
        public void SortNodes(Span<Node> nodes)
        {
            var sortList = new SortBucketList?[_maxN + 1];

            for (int i = 0; i < nodes.Length; i++)
            {
                var node = nodes[i];
                var pairsCount = _path == null ? node.PairsCount : PairsNotInPath(node, _path);

                var list = sortList[pairsCount];
                if (list == null)
                {
                    list = new SortBucketList(_path);
                }
                
                list.AddNode(node);

                sortList[pairsCount] = list;
            }

            var targetPos = 0;
            for (var i = 0; i < sortList.Length; i++)
            {
                var list = sortList[i];
                if (list == null)
                {
                    continue;
                }

                for (var j = 0; j < list.Nodes.Count; j++)
                {
                    var node = list.Nodes[j];
                    nodes[targetPos] = node;
                    targetPos++;
                }
            }
        }
        
        private static int PairsNotInPath(Node node, Path path)
        {
            var count = 0;
            var pairs = node.Pairs;
            for (var i = 0; i < pairs.Length; i++)
            {
                if (!path.Contains(pairs[i].Value))
                {
                    count++;
                }
            }

            return count;
        }
    }
}